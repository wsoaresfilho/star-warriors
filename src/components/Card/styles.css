.cards-container {
  display: flex;
  width: 800px;
  flex-direction: row;
  overflow: auto;
  flex-wrap: wrap;
  margin: auto;
  height: calc(100vh - 220px);
}

.card__container {
  width: calc(100% / 2 - 20px);
  margin: 10px;
}

.card__content {
  padding: 32px 5px;
}

.card__image {
  width: 100%;
  height: 300px;
  background-color: gray;
  overflow: hidden;
  display: flex;
}

.card__image img {
  max-width: 100%;
  width: auto;
  min-height: 100%;
  height: auto;
  margin: auto;
  object-fit: cover;
}

.loading {
  width: 100%;
  height: calc(100vh - 220px);
  text-align: center;
  position: relative;
  margin-left: -20px;
  margin-top: -20px;
}

.loading img {
  max-width: 40px;
  position: absolute;
  top: 50%;
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.cards-container::-webkit-scrollbar-track {
  -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  background-color: var(--light-bg-color);
}

.dark .cards-container::-webkit-scrollbar-track {
  -webkit-box-shadow: inset 0 0 6px rgba(149, 149, 149, 0.3);
  box-shadow: inset 0 0 6px rgba(149, 149, 149, 0.3);
  background-color: var(--dark-bg-color);
}

.cards-container::-webkit-scrollbar {
  width: 7px;
}

.cards-container::-webkit-scrollbar-thumb {
  background-color: var(--dark-bg-color);
}

.dark .cards-container::-webkit-scrollbar-thumb {
  background-color: var(--light-bg-color);
}
