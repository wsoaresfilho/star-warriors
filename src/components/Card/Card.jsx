import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import CustomSwitch from '../CustomSwitch/CustomSwitch';
import './styles.css';

class Card extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            isInPersonalPanel: true,
        };

        this.handlePanelChange = this.handlePanelChange.bind(this);
    }

    handlePanelChange() {
        const { isInPersonalPanel } = this.state;
        this.setState({
            isInPersonalPanel: !isInPersonalPanel,
        });
    }

    render() {
        const {
            name,
            imgUrl,
            birth_year,
            gender,
            height,
            mass,
            starshipsDetails,
        } = this.props;
        const { isInPersonalPanel } = this.state;
        const hasStarships = starshipsDetails.length > 0;
        return (
            <div className='card-container'>
                <div className='card-container__image'>
                    <img src={imgUrl} alt={name} />
                </div>
                <div className='card-content'>
                    <h1 className='card-content__title'>{name}</h1>
                    {isInPersonalPanel && (
                        <div className='card-content__panel personal'>
                            <div className='card-content__info'>{`Gender: ${gender}`}</div>
                            <div className='card-content__info'>{`Birth Year: ${birth_year}`}</div>
                            <div className='card-content__info'>{`Height: ${height} cm`}</div>
                            <div className='card-content__info'>{`Weight: ${mass} Kg`}</div>
                        </div>
                    )}
                    {!isInPersonalPanel && (
                        <div className='card-content__panel'>
                            {!hasStarships && (
                                <div className='card-content__panel-nostarships'>
                                    There are no starships for this character
                                </div>
                            )}
                            {hasStarships &&
                                starshipsDetails.map(starship => {
                                    return (
                                        <div
                                            key={starship.name}
                                            className='card-content__panel-starships'
                                        >
                                            <div className='card-content__info'>
                                                {starship.name}
                                            </div>
                                            <div className='card-content__info'>{`Model: ${starship.model}`}</div>
                                            <div className='card-content__info'>{`Class: ${starship.starship_class}`}</div>
                                            <div className='card-content__info'>{`Num passengers: ${starship.passengers}`}</div>
                                            <div className='card-content__info'>{`Length: ${starship.length}`}</div>
                                        </div>
                                    );
                                })}
                        </div>
                    )}
                    <div className='card-content__panel-switch'>
                        <CustomSwitch
                            onChange={this.handlePanelChange}
                            isChecked={!isInPersonalPanel}
                            isSmall
                            leftText='Personal'
                            rightText='Starships'
                        />
                    </div>
                </div>
            </div>
        );
    }
}

Card.propTypes = {
    name: PropTypes.string.isRequired,
    imgUrl: PropTypes.string.isRequired,
    birth_year: PropTypes.string.isRequired,
    gender: PropTypes.string.isRequired,
    height: PropTypes.string.isRequired,
    mass: PropTypes.string.isRequired,
    starshipsDetails: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string,
            model: PropTypes.string,
            starship_class: PropTypes.string,
            passengers: PropTypes.string,
            length: PropTypes.string,
        })
    ).isRequired,
};

export default Card;
