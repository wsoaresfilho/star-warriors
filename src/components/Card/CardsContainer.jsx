import React, { PureComponent } from 'react';
import Card from './Card';
import { getDataFromApi, getImageByTerm } from '../../api/api';
import getIntFromString from '../../utils/helpers';

class CardsContainer extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            cards: [],
            isFetching: false,
            next: null,
            addedCardsNum: 0,
            starshipsMap: new Map(),
        };
    }

    componentDidMount() {
        this.fetchCards();
    }

    componentDidUpdate() {
        const { cards, isFetching } = this.state;
        if (cards.length > 0 && !isFetching) {
            this.startLazyLoadingCards();
        }
    }

    getLoadingElement() {
        const { isFetching } = this.state;
        return (
            <div className='loading'>
                {isFetching && <img src='loading.gif' alt='Loading' />}
            </div>
        );
    }

    async fetchCards() {
        this.setState({ isFetching: true });

        const { next: link, cards, starshipsMap } = this.state;
        const people = await getDataFromApi(link);
        const { results, next } = people;

        if (results.length > 0) {
            const updatedResults = await Promise.all(
                results.map(async res => {
                    const imageResp = await getImageByTerm(res.name);
                    const imageUrl =
                        imageResp.items && imageResp.items.length > 0
                            ? imageResp.items[0].link
                            : '';
                    const { starships } = res;

                    const starshipsDetails = await Promise.all(
                        starships.map(async sts => {
                            const starShipId = getIntFromString(sts);
                            const stsMap = starshipsMap.has(starShipId);
                            if (stsMap) {
                                return starshipsMap.get(starShipId);
                            }

                            const newStarship = await getDataFromApi(sts);
                            this.setState({
                                starshipsMap: starshipsMap.set(
                                    starShipId,
                                    newStarship
                                ),
                            });
                            return newStarship;
                        })
                    );

                    const updatedRes = {
                        ...res,
                        imgUrl: imageUrl,
                        starshipsDetails,
                    };
                    return updatedRes;
                })
            );

            this.setState({
                cards: cards.concat(updatedResults),
                isFetching: false,
                next,
                addedCardsNum: results.length,
            });
        }
    }

    startLazyLoadingCards() {
        const { next, addedCardsNum } = this.state;
        const elements = [].slice.call(
            document.querySelectorAll('.cards-container img')
        );

        if ('IntersectionObserver' in window) {
            const lazyImageObserver = new IntersectionObserver(entries => {
                entries.forEach(async entry => {
                    if (entry.isIntersecting) {
                        const lazyImage = entry.target;
                        lazyImageObserver.unobserve(lazyImage);

                        if (next) {
                            this.fetchCards();
                        }
                    }
                });
            });

            elements.slice(-addedCardsNum).forEach((lazyImage, index) => {
                if ((index + 1) % 7 === 0) {
                    lazyImageObserver.observe(lazyImage);
                }
            });
        }
    }

    render() {
        const { cards } = this.state;
        return (
            <div className='cards-main'>
                <h1 className='cards-container__title'>Star Wars Characters</h1>
                <div className='cards-container'>
                    {cards.map(card => {
                        const props = card;
                        return <Card key={card.name} {...props} />;
                    })}
                    {this.getLoadingElement()}
                </div>
            </div>
        );
    }
}

export default CardsContainer;
