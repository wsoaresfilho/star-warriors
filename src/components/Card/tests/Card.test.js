import React from 'react';
import { mount } from 'enzyme';
import Card from '../Card';

describe('Card component', () => {
    let wrapper;
    let instance;

    const defaultProps = {
        name: 'John',
        imgUrl: 'http://teste.com/teste.png',
        birth_year: '1980',
        gender: 'male',
        height: '188',
        mass: '100',
        starshipsDetails: [
            {
                name: 'star',
                model: 'big model',
                starship_class: 'class1',
                passengers: '10',
                length: '20',
            },
            {
                name: 'star2',
                model: 'small model',
                starship_class: 'class2',
                passengers: '2',
                length: '10',
            },
        ],
    };

    beforeEach(() => {
        wrapper = mount(<Card {...defaultProps} />);
        instance = wrapper.instance();
    });

    afterEach(() => {
        wrapper.unmount();
    });

    describe('when initializes should render', () => {
        it('card-container element', () => {
            expect(wrapper.find('.card-container').length).toBe(1);
        });

        it('card-container title', () => {
            expect(wrapper.find('.card-content__title').length).toBe(1);
            expect(wrapper.find('.card-content__title').text()).toBe('John');
        });

        it('card image', () => {
            expect(wrapper.find('.card-container__image').length).toBe(1);
            expect(wrapper.find('.card-container__image img').length).toBe(1);
        });

        it('personal panel', () => {
            expect(wrapper.find('.card-content__panel.personal').length).toBe(
                1
            );
        });

        it('not show starship panel', () => {
            expect(wrapper.find('.card-content__panel-starships').length).toBe(
                0
            );
        });

        it('CustomSwitch component', () => {
            expect(wrapper.find('CustomSwitch').length).toBe(1);
        });

        it('snapshot correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when changes to starship panel', () => {
        it('should render starship panel', () => {
            instance.setState({ isInPersonalPanel: false });
            wrapper.update();

            expect(wrapper.find('.card-content__panel-starships').length).toBe(
                2
            );
        });

        it('should not render starship panel when there are no starships', () => {
            const newProps = {
                ...defaultProps,
                starshipsDetails: [],
            };
            wrapper = mount(<Card {...newProps} />);
            wrapper.instance().setState({ isInPersonalPanel: false });
            wrapper.update();

            expect(
                wrapper.find('.card-content__panel-nostarships').text()
            ).toBe('There are no starships for this character');
        });

        it('snapshot correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
