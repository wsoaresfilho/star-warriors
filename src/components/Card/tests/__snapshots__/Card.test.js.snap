// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component when changes to starship panel snapshot correctly 1`] = `ReactWrapper {}`;

exports[`Card component when initializes should render snapshot correctly 1`] = `ReactWrapper {}`;
